python/pyspark/sql/streaming/python_streaming_source_runner.py

```diff
@@ -88,25 +94,86 @@ def partitions_func(
         if it is None:
             write_int(PREFETCHED_RECORDS_NOT_FOUND, outfile)
         else:
-            send_batch_func(it, outfile, schema, max_arrow_batch_size, data_source)
+            send_batch_func(it, outfile, schema, max_arrow_batch_size, data_source)  # noqa: E501
     else:
         write_int(PREFETCHED_RECORDS_NOT_FOUND, outfile)
 
 
-def commit_func(reader: DataSourceStreamReader, infile: IO, outfile: IO) -> None:
+def commit_func(reader: DataSourceStreamReader, infile: IO, outfile: IO) -> None:  # noqa: E501
     end_offset = json.loads(utf8_deserializer.loads(infile))
     reader.commit(end_offset)
     write_int(0, outfile)
 
 
+def latest_offset_with_report_func(reader: DataSourceStreamReader, infile: IO, outfile: IO) -> None:
+    """
+    Handler for function ID 890: latestOffset with admission control
+    parameters.
+
+    This function supports both old and new reader implementations:
+    - Old readers: latestOffset() with no parameters -> no admission
+      control
+    - New readers: latestOffset(start, limit) with parameters ->
+      admission control enabled
+    """
+    start_offset_str = utf8_deserializer.loads(infile)
+    # Handle empty string as None for backward compatibility
+    start_offset = json.loads(start_offset_str) if start_offset_str else None
+    limit = json.loads(utf8_deserializer.loads(infile))
+
+    # Type declarations for mypy
+    capped_offset: dict
+    true_latest_offset: dict
+
+    try:
+        # Try calling with optional parameters (new signature)
+        result = reader.latestOffset(start_offset, limit)
+
+        # Check return type to determine behavior
+        if isinstance(result, tuple):
+            # New behavior: returns (capped_offset, true_latest_offset)
+            capped_offset, true_latest_offset = result
+        else:
+            # Old behavior or no admission control: single offset
+            capped_offset = true_latest_offset = result
+
+    except TypeError:
+        # Old signature that doesn't accept parameters - fallback
+        try:
+            fallback_result = reader.latestOffset()
+            # Handle both return types for backward compatibility
+            if isinstance(fallback_result, tuple):
+                capped_offset, true_latest_offset = fallback_result
+            else:
+                capped_offset = true_latest_offset = fallback_result
+        except Exception as fallback_error:
+            raise IllegalArgumentException(
```

Comment:
I'm not very sure this is a representative exception. Shall we not wrap the exception and throw it as it is? If you want to give the info of operation into exception, it's OK to wrap but IllegalArgumentException is too narrow one to cover "Exception".